index.src.html

```diff
@@ -2394,14 +2420,692 @@
+
+        : [=read request/chunk steps=], given |chunk|
+        :: 1. If {{ImageDecoder/[[fatal error]]}} is `true`, abort these steps.
+            2. If |chunk| is not a Uint8Array object, run the [=Handle Error=]
+                algorithm with a {{NotReadableError}} {{DOMException}} and
+                abort these steps.
+            3. Let |bytes| be the byte sequence represented by the Uint8Array
+                object.
+            4. Append |bytes| to the  {{ImageDecoder/[[encoded data]]}}
+                internal slot.
+            5. If {{ImageDecoder/[[tracks established]]}} is `false`, run the
+                [=Establish Tracks=] algorithm.
+            6. Otherwise, run the [=Update Tracks=] algorithm.
+            7. Run the [=Fetch Stream Data Loop=] algorithm with |reader|.
+
+        : [=read request/close steps=]
+        :: 1. Assign `true` to {{ImageDecoder/complete}}
+
+        : [=read request/error steps=]
+        :: 1. Queue a task on the [=control thread=] event loop to run the
+                [=Handle Error=] algorithm.
+
+    2. Read a chunk from |reader| given |readRequest|.
+
+: <dfn for=ImageDecoder>Establish Tracks</dfn>
+:: Run these steps:
+    1. Assert {{ImageDecoder/[[tracks established]]}} is `false`.
+    2. If {{ImageDecoder/[[encoded data]]}} does not contain enough data to
+        determine the number of tracks, abort these steps.
+    3. If the number of tracks is found to be '0', run the
+        [=ImageDecoder/Handle Error=] algorithm and abort these steps.
+    4. Let |selectedTrackIndex| be `-1`.
+    5. Let |newTrackList| be a new [=list=].
+    6. For each |image track| found in {{ImageDecoder/[[encoded data]]}}:
+        1. Let |newTrack| be a new {{ImageTrack}}, initialized as follows:
+            1. Assign [=this=] to {{ImageTrack/[[ImageDecoder]]}}.
+            2. Assign {{ImageDecoder/tracks}} to
+                {{ImageTrack/[[ImageTrackList]]}}.
+            3. If |image track| is found to be animated, assign `true` to
+                |newTrack|'s {{ImageTrack/[[animated]]}} internal slot.
+                Otherwise, assign `false`.
+            4. If |image track| is found to describe a frame count, assign
+                that count to |newTrack|'s {{ImageTrack/[[frame count]]}}
+                internal slot. Otherwise, assign `0`.
+
+                NOTE: If [=this=] was constructed with
+                  {{ImageDecoderInit/data}} as a {{ReadableStream}}, the
+                  {{ImageTrack/frameCount}} may change as additional bytes are
+                  appended to {{ImageDecoder/[[encoded data]]}}. See the
+                  [=Update Tracks=] algorithm.
+
+            5. If |image track| is found to describe a repetition count,
+                assign that count to {{ImageTrack/[[repetition count]]}}
+                internal slot. Otherwise, assign `0`.
+
+                NOTE: A value of `Infinity` indicates infinite repetitions.
```

Comment:
Pulling in your comment here: https://bugs.chromium.org/p/chromium/issues/detail?id=1183456#c2

> To handle infinity we can either use zero like I do today or make the repetitionCount an optional field which isn't present when the value == not specified or infinity.

Overloading zero isn't great since zero is also a valid repetition count. IDL-wise, I'm only able to make dictionary members optional (ImageTrack is an interface). For interfaces the closest is probably nullable (`unsigned long? repetitionCount`). But I'm not in love with "null means zero, and zero means infinity". Do you like that better than float repetitionCount? I appreciate that it's weird to use float purely for the sake of including Infinity, but in practice I don't think this will ever cause anyone pain.

Reply:
Is zero valid? I think we should just require once to have a value of 1 and define anything other than that as zero. I think it's obvious that animated = true and repetitionCount = 0, means up to you.

Reply:
To me zero implies "play once through". If the name were playCount rather than repetitionCount, I could see your meaning.

Reply:
Let's just call playCount then? Maybe make it optional if it's infinity or not specified?

Reply:
I.e., let's go with simple meaning versus exporting complexity